kernel/src/fs/cgroupfs/systree_node.rs

```
@@ -125,45 +339,157 @@ inherit_sys_branch_node!(CgroupSystem, fields, {
+                    writeln!(printer, "{}", pid)?;
+                }
+            }
+            "cgroup.events" => {
+                let res = if self.populated_count.load(Ordering::Relaxed) > 0 {
+                    1
+                } else {
+                    0
+                };
+
+                writeln!(printer, "populated {}", res)?;
+                // Currently we have not enabled the "frozen" attribute
+                // so the "frozen" field is always zero.
+                writeln!(printer, "frozen {}", 0)?;
+            }
+            _ => {
+                // TODO: Add support for reading other attributes.
+                return Err(Error::AttributeError);
+            }
+        }
+
+        Ok(printer.bytes_written())
     }
 
-    fn write_attr(&self, _name: &str, _reader: &mut VmReader) -> Result<usize> {
-        // TODO: Add support for writing attributes.
-        Err(Error::AttributeError)
+    fn write_attr(&self, name: &str, reader: &mut VmReader) -> Result<usize> {
+        match name {
+            "cgroup.procs" => {
+                let (content, len) = reader
+                    .read_cstring_until_end(MAX_ATTR_SIZE)
+                    .map_err(|_| Error::PageFault)?;
+                let pid = content
+                    .to_str()
+                    .ok()
+                    .and_then(|string| string.trim().parse::<Pid>().ok())
+                    .ok_or(Error::InvalidOperation)?;
+
+                with_process_cgroup_locked(pid, |process, cgroup_membership| {
+                    // TODO: According to the "no internal processes" rule of cgroupv2
+                    // (Ref: https://man7.org/linux/man-pages/man7/cgroups.7.html),
+                    // if the cgroup node has enabled some controllers like "memory", "io",
+                    // it is forbidden to bind a process to an internal cgroup node.
+                    cgroup_membership.move_process_to_node(process, self);
```

What will happen and what should happen, if we perform something like this?
```c
fd = open("/sys/fs/cgroup/test/cgroup.procs", O_WRONLY);
rmdir("/sys/fs/cgroup/test");
write(fd, "1234", 4);
```
```c
fd = open("/sys/fs/cgroup/foo/", O_PATH);
rmdir("/sys/fs/cgroup/foo");
mkdirat(fd, "bar");
```

I guess we need to return an error.

Instead of having a `try_run_if_empty`, perhaps we need something like this (which seems to avoid the race condition I mentioned previously):
```rust
        let child = branch_node.get_child(name)?;

        // This will succeed only if the child is empty and has not been removed.
        child.mark_as_removed()?;

        // The child can be removed at most once. No other threads can remove it concurrently.
        branch_node.remove_child(name).unwrap();
```

`mark_as_removed` sets a flag. `move_process_to_node` and `create_child` must fail if the flag is set.